drivers/accel/adxl38x/adxl38x.c

```diff
@@ -0,0 +1,957 @@
+{
+	struct adxl38x_dev *dev;
+	int ret;
+	uint8_t reg_value;
+
+	dev = (struct adxl38x_dev *)no_os_calloc(1, sizeof(*dev));
+	if (!dev)
+		return -ENOMEM;
+
+	switch (init_param.dev_type) {
+	case ID_ADXL380:
+	case ID_ADXL382:
+		dev->dev_type = init_param.dev_type;
+		break;
+	default:
+		return -EINVAL;
+	}
+
+	dev->comm_type = init_param.comm_type;
+	if (dev->comm_type == ADXL38X_SPI_COMM) {
+		ret = no_os_spi_init(&dev->com_desc.spi_desc, &init_param.comm_init.spi_init);
+		if (ret)
+			goto error_dev;
+	} else {
+		ret = no_os_i2c_init(&dev->com_desc.i2c_desc, &init_param.comm_init.i2c_init);
+		if (ret)
+			goto error_dev;
+	}
+
+	ret = adxl38x_read_device_data(dev, ADXL38X_DEVID_AD,
+				       1, &reg_value);
+	if (ret || (reg_value != ADXL38X_RESET_DEVID_AD))
+		goto error_com;
+
+	ret = adxl38x_read_device_data(dev, ADXL38X_DEVID_MST,
+				       1, &reg_value);
+	if (ret || (reg_value != ADXL38X_RESET_DEVID_MST))
+		goto error_com;
+
+	ret = adxl38x_read_device_data(dev, ADXL38X_PART_ID,
+				       1, &reg_value);
+	if (ret || (reg_value != ADXL38X_RESET_PART_ID))
+		goto error_com;
+
+	/* Set device type 380/382 */
+	ret = adxl38x_read_device_data(dev, ADXL38X_MISC0,
+				       1, &reg_value);
+	if (ret)
+		goto error_com;
+	if ((reg_value & NO_OS_BIT(7))) {
+		dev->dev_type = ID_ADXL380;
+	} else {
+		dev->dev_type = ID_ADXL382;
+	}
+
+	/* Set default values to the device structure */
+	dev->range = ADXL380_RANGE_4G;
+	dev->op_mode = ADXL38X_MODE_STDBY;
+	*device = dev;
+
+	return 0;
+
+error_com:
+	if (dev->comm_type == ADXL38X_SPI_COMM)
+		no_os_spi_remove(dev->com_desc.spi_desc);
+	else
+		no_os_i2c_remove(dev->com_desc.i2c_desc);
+	no_os_free(dev);
+	return -1;
+error_dev:
+	no_os_free(dev);
+	return ret;
+}
+
+/***************************************************************************//**
+ * @brief Free the resources allocated by the init function.
+ *
+ * @param dev  - The device structure.
+ *
+ * @return ret - Result of the remove procedure.
+*******************************************************************************/
+int adxl38x_remove(struct adxl38x_dev *dev)
+{
+	int ret;
+
+	if (dev->comm_type == ADXL38X_SPI_COMM)
+		ret = no_os_spi_remove(dev->com_desc.spi_desc);
+	else
+		ret = no_os_i2c_remove(dev->com_desc.i2c_desc);
+	no_os_free(dev);
+
+	return ret;
+}
+
+/***************************************************************************//**
+ * @brief Performs a soft reset of the device.
+ *
+ * @param dev  - The device structure.
+ *
+ * @return ret - Result of the soft reset procedure.
+*******************************************************************************/
+int adxl38x_soft_reset(struct adxl38x_dev *dev)
+{
+	uint8_t reg_value;
+	int ret;
+	uint8_t data = ADXL38X_RESET_CODE;
+
+	// Perform soft reset
+	ret = adxl38x_write_device_data(dev, ADXL38X_REG_RESET, 1, &data);
+	if (ret)
+		return ret;
+	// Delay is needed between soft reset and initialization
+	no_os_udelay(1000);
```

Comment:
if this delay is needed it's documented in the datasheet, we try to specify that here (page number, section, or name of the time constant like Treset or whatever it's called) 

this shows the value is not empirical but well documented